File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/FamilyBookViewController.m
//
//  FamilyBookViewController.m
//  
//
//  Created by Federal Authority For Identity and Citizenship on 5/14/17.
//  Copyright © 2017 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "FamilyBookViewController.h"

@interface FamilyBookViewController () {
    
    NSMutableArray *pHeadofFamilyKeyarr;
    NSMutableArray *pHeadofFamilyValuearr;
}
@property(nonatomic,strong)Utils * utils;
@end

@implementation FamilyBookViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
     self.listTable.hidden=YES;
     self.familyBookEnterPinView.hidden=NO;
    
        // init objects
    pHeadofFamilyKeyarr =[[NSMutableArray alloc]init];
    pHeadofFamilyValuearr=[[NSMutableArray alloc]init];
    
    self.pintext.keyboardType=UIKeyboardTypeNumberPad;
    self.pintext.keyboardAppearance = UIKeyboardAppearanceLight;
    UIToolbar* enterPinNumberToolbar = [[UIToolbar alloc]initWithFrame:CGRectMake(0, 0, 320, 50)];
    enterPinNumberToolbar.barStyle = UIBarStyleBlackTranslucent;
    enterPinNumberToolbar.items = @[[[UIBarButtonItem alloc]initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinCancelNumberPad)],
                                    [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                                    [[UIBarButtonItem alloc]initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinDoneWithNumberPad)]];
    [enterPinNumberToolbar sizeToFit];
    self.pintext.inputAccessoryView = enterPinNumberToolbar;
    
    self.title=@"Family book";
    self.navigationItem.hidesBackButton = YES;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(goBack)];
    
    self.readfamilyBookButton.layer.cornerRadius=22.0;
    [self.readfamilyBookButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.readfamilyBookButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.readfamilyBookButton.layer setShadowOpacity:0.5];
   // self.readfamilyBookButton.layer.borderWidth=1.0;
    
    AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    self.model = aDelegate.model;
    self.utils=aDelegate.utils;
    
    // Do any additional setup after loading the view.
}
#pragma mark viewWillAppear

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:YES];
    
    self.pintext.text=@"";
    [self.pintext resignFirstResponder];
}
- (IBAction)readfamilyBookButtonAction:(id)sender {
    
    
    [self.pintext resignFirstResponder];
    if ([self.pintext.text length]==0) {
        [AlertView showAlertTitle:ALERT withMessage:ENTER_PIN onView:self];
    }// if
    else {
    if ([self.pintext.text length]<4) {
        [AlertView showAlertTitle:ALERT withMessage:MIN_PIN onView:self];
        self.pintext.text=@"";
    }// if
    else if ([self.pintext.text length]>16) {
        [AlertView showAlertTitle:ALERT withMessage:MAX_PIN onView:self];
        self.pintext.text=@"";
    }// else if
    else {
      [self.utils ShowProgressBar:READING_CARDDEATILS andView:self.view];
      NSString *requestId=[Utils generateSecureKey];
      dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
      @try {          
          CardReader *cardreader  = [self.model.cardReaderSharedArray objectAtIndex:0];
          DataProtectionKey *data_protection_key=self.model.dataprotectedShared;
          
        NSString *request_handle =  [cardreader prepareRequest:requestId];
        if (request_handle.length==0 || [request_handle isEqualToString:@""] || [request_handle isEqualToString:@"(null)"] ) {
            [AlertView showAlertTitle:ALERT withMessage:REQUEST_HANDLE_EMPTY onView:self];
        }//if
        else {
            NSString *encodePin=[Utils setEncrytion:request_handle data:self.pintext.text publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
            if (encodePin.length==0 || [encodePin isEqualToString:@""] || [encodePin isEqualToString:@"(null)"] ) {
                [AlertView showAlertTitle:ALERT withMessage:ENCODE_PIN_EMPTY onView:self];
            }//if
            else {
            CardFamilyBookData *familybook = [cardreader readFamilyBookData:encodePin];
            dispatch_async(dispatch_get_main_queue(), ^{
                NSString *xmlString =[familybook getXmlString];
                NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:xmlString];
                if (![errorMsg isEqualToString:@""] || errorMsg.length>0) {
                    [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                }//if
                else {
                    FamilyBookParse *FamilyBookparse=[[FamilyBookParse alloc]init];
                    HeadOfFamily *headfamily = [familybook getHeadOfFamily];
                    [FamilyBookparse getHeadOfFamilyDetails:headfamily];
                    
                    NSArray *wifeArr=[familybook getWives];
                    for (int i =0; i<wifeArr.count; i++) {
                        Wife *wife = (Wife *)[wifeArr objectAtIndex:i];
                        [FamilyBookparse getWifeDetails:wife index:i];
                    }//for loop
                        
                    NSArray *childArr=[familybook getChildren];
                    for (int i =0; i<childArr.count; i++) {
                        Child *child = (Child *)[childArr objectAtIndex:i];
                        [FamilyBookparse getChildDetails:child index:i];
                    }//for loop
                    [self->pHeadofFamilyKeyarr addObjectsFromArray:[FamilyBookparse familyBookKey]];
                    [self->pHeadofFamilyValuearr addObjectsFromArray:[FamilyBookparse familyBookValue]];
                }//else
               });// update UI main queue
             }//else
           }//else
         }//try
          @catch (NSException *exception) {
              NSString *vgresponse =exception.userInfo[@"ErrorResponse"];
              if (![vgresponse isEqualToString:@""] && vgresponse.length>0) {
                  NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:vgresponse];
                  if (![errorMsg isEqualToString:@""] && errorMsg.length>0) {
                      [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                  }//if
                  else {
                      NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                      [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                  }//else
              }//if
              else {
                  NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                  [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
              }
          }// catch
        @finally {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.utils DismissProgressBar];
                self.pintext.text=@"";
                if (self->pHeadofFamilyKeyarr.count>0) {
                    self.familyBookEnterPinView.hidden=YES;
                    self.listTable.hidden=NO;
                    [self.listTable reloadData];
                }// if
             });// update UI main queue
           }// finally
         });//background queue
       }//else
   }//else
}
#pragma mark Updating UI Methods

-(void)enterPinCancelNumberPad {
    [self.pintext resignFirstResponder];
    self.pintext.text = @"";
}
-(void)enterPinDoneWithNumberPad {
    [self.pintext resignFirstResponder];
}
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [self.pintext setUserInteractionEnabled:YES];
    [self.pintext resignFirstResponder];
    return YES;
}
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
#pragma mark TableView Delegate and Data source Methods

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
     return 1;;
}
-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section{
    return @"Family Book Data";
}
- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section {
    view.tintColor = [Utils colorFromHexString:@"#F4F0E8"];
    
}
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section; {
    return 32.0;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return pHeadofFamilyKeyarr.count;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *customTabelId=@"identifierCell";
    CustomTableViewCell *cell=(CustomTableViewCell *)[tableView dequeueReusableCellWithIdentifier:customTabelId];
    if (cell == nil) {
        cell = [[CustomTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:customTabelId];
    }// if
    cell.keyData.text = [NSString stringWithFormat:@"%@",[pHeadofFamilyKeyarr objectAtIndex:indexPath.row]];
    cell.valueData.text = [NSString stringWithFormat:@"%@",[pHeadofFamilyValuearr objectAtIndex:indexPath.row]];
    
    return cell;
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */





@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/FamilyBookViewController.swift
//
//  FamilyBookViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 25/04/22.
//

import UIKit

class FamilyBookViewController: UIViewController,UITableViewDelegate, UITableViewDataSource {
    
    @IBOutlet weak var pintext: UITextField!
    @IBOutlet var listTable: UITableView!
    @IBOutlet var readfamilyBookButton: UIButton!
    
    @IBOutlet weak var view_Roundedcorners: UIView!
    
    var pHeadofFamilyKeyarr = NSMutableArray()
    var pHeadofFamilyValuearr = NSMutableArray()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        listTable.isHidden = true
        
        self.pintext.keyboardType = .numberPad
        self.pintext.keyboardAppearance = .light
        let enterPinNumberToolbar = UIToolbar(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(320), height: CGFloat(50)))
        enterPinNumberToolbar.barStyle = .blackTranslucent
        enterPinNumberToolbar.items = [UIBarButtonItem(title: "Cancel", style: .done, target: self, action: #selector(enterPinCancelNumberPad)), UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(enterPinDoneWithNumberPad))]
        enterPinNumberToolbar.sizeToFit()
        self.pintext.inputAccessoryView=enterPinNumberToolbar
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)

        // Do any additional setup after loading the view.
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        pintext.text=""
        pintext.resignFirstResponder()
    }
    
    @IBAction func readfamilyBookButtonAction(_ sender: Any) {
        
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        else if cardreader == nil {
            self.view.makeToast(Constants.cardreaderNotConnect)
            return
        }//else if
        
        self.pintext.resignFirstResponder()
        if self.pintext.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterPin, onView: self)
        }// if
        else {
            if (self.pintext.text?.count)!<4 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MinPin, onView: self)
                self.pintext.text=""
            }// if
            else if (self.pintext.text?.count)!>16 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MaxPin, onView: self)
                self.pintext.text=""
            }// else if
            else {
                utils.showProgressBar(Constants.ReadingCardDetails, andView: self.view)
                disableTheUIelements()
                let requestId = Utils.generateSecureKey()
                DispatchQueue.global(qos: .background).async {
                    do {
                        let request_handle = try cardreader?.prepareRequest(requestId!)
                        if (request_handle?.isEmpty == true) || (request_handle == "") || (request_handle == "(null)") {
                            AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Request_Handle_Empty, onView: self)
                        }//if
                        else {
                            
                            //                 let encodePin:String? = Utils.setEncrytion(request_handle, data: self.pintext.tex)
                            let encodePin:String? = Utils.setEncrytion(request_handle, data: self.pintext.text!, publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()) )
                            if (encodePin?.isEmpty)! || (encodePin == "") || (encodePin == "(null)") {
                                AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_Pin_Empty, onView: self)
                            }//if
                            else {
                                familybook = try cardreader?.readFamilyBookData(encodePin!)
                                DispatchQueue.main.async {
                                    let xmlString = familybook.getXmlString()
                                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: xmlString)
                                    if (errorMsg != "") && (errorMsg.count ) > 0 {
                                        AlertView.showAlertTitle(Constants.Alert, withMessage: errorMsg, onView: self)
                                    }//if
                                    else {
                                        headfamily = familybook.getHeadOfFamily()
                                        let FamilyBookparse = FamilyBookParse()
                                        FamilyBookparse.getHeadOfFamilyDetails(headfamily!)
                                        
                                        var wifeArr = NSArray()
                                        wifeArr = familybook.getWives()
                                        for i in 0..<wifeArr.count {
                                            wifedata = wifeArr[i] as? Wife
                                            FamilyBookparse.getWifeDetails(wifedata!, index: i)
                                        }//for
                                        var childArr = NSArray()
                                        childArr = familybook.getChildren()
                                        for i in 0..<childArr.count {
                                            childdata = childArr[i] as? Child
                                            FamilyBookparse.getChildDetails(childdata!, index: i)
                                        }//for
                                        self.pHeadofFamilyKeyarr = FamilyBookparse.familyBookKey()
                                        self.pHeadofFamilyValuearr = FamilyBookparse.familyBookValue()
                                    }//else
                                    if self.pHeadofFamilyKeyarr.count > 0 {
                                        self.pintext.text=""
                                       
                                        self.listTable.isHidden=false
                                        self.listTable.reloadData()
                                    }// if
                                }// update UI main queue
                                DispatchQueue.main.async {
                                    utils.dismissProgressBar()
                                    self.pintext.text=""
                                }// update UI main queue
                            }//else
                        }//else
                    }//do
                    catch let error as NSError{
                        DispatchQueue.main.async {
                            var vgresponse = String()
                            if error.userInfo["ErrorResponse"] != nil {
                                vgresponse = error.userInfo["ErrorResponse"] as! String
                                if (vgresponse != "") && ((vgresponse.count ) > 0) {
                                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: vgresponse)
                                    if (errorMsg != "") && ((errorMsg.count ) > 0) {
                                        AlertView.showAlertTitle(Constants.Alert, withMessage:errorMsg, onView: self)
                                    }//if
                                    else {
                                        let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                        AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                                    }//else
                                }//if
                                else {
                                    let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                    AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                                }//else
                            }//if
                            else {
                                let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                                AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                            }//else
                            
                            utils.dismissProgressBar()
                            self.pintext.text=""
                        }// update UI main queue
                    }//catch
                }//background queue
            }//else
        }//else
    }
    

    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    
   
    @objc func enterPinCancelNumberPad() {
        pintext.resignFirstResponder()
        pintext.text=""
    }
    
    @objc func enterPinDoneWithNumberPad() {
        pintext.resignFirstResponder()
    }
    
    func disableTheUIelements() {
        pHeadofFamilyKeyarr.removeAllObjects()
        pHeadofFamilyValuearr.removeAllObjects()
    }
    
    // MARK:TableView Delegate and Data source Methods
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return pHeadofFamilyKeyarr.count
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "HeadofFamily"
    }
    
    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        view.tintColor = Utils.color(fromHexString:"#F4F0E8")
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 32.0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let customTabelId="identifierCell"
        let cell: CustomTableViewCell=(tableView.dequeueReusableCell(withIdentifier: customTabelId) as? CustomTableViewCell)!
        cell.keyData?.text = pHeadofFamilyKeyarr[indexPath.row] as? String
        cell.valueData?.text = pHeadofFamilyValuearr[indexPath.row] as? String
        
        return cell
    }
}
